[Modularization] add possibility to extend or overwrite RPC calls

jonasschnelli opened the pull request:
This will allow a upcomming modularization of the wallet or other components.
Currently there is no use of the flexible RPC dispatch table.

Because #5686 is getting to big, it might be a better approach to slice out independent changes which can be easily reviewed and tested.

jonasschnelli committed:
[Modularization] add possibility to extend or overwrite RPC calls

- this will allow a upcomming modularization of the wallet or other components

jtimon commented:
ut ACK

sipa commented on src/rpcserver.cpp:
```diff
@@ -368,6 +368,16 @@ const CRPCCommand *CRPCTable::operator[](string name) const
+void CRPCTable::AddOrReplaceCommand(const CRPCCommand command)
+{
+    // search after existing key in hashmap and remove it
+    map<string, const CRPCCommand*>::iterator it = mapCommands.find(command.name);
```
Why is this lookup needed? You can just overwrite using the line you have below?

jonasschnelli replied:
Hmm... yes. Should be sufficient to just overwrite. Will change this.

sipa commented on src/rpcserver.cpp:
```diff
@@ -368,6 +368,16 @@ const CRPCCommand *CRPCTable::operator[](string name) const
 
+void CRPCTable::AddOrReplaceCommand(const CRPCCommand command)
+{
+    // search after existing key in hashmap and remove it
```
map is also not a hashmap, but a treemap (but that shouldn't really matter).

dexX7 commented:
Hey @jonasschnelli,

this is a very interesting idea. I faced some issues (segfaults and std::bad_alloc exceptions), probably related to object lifetime, and ended up passing `const CRPCCommand* command`, where the pointer is maintaned externally, which seemed more intuitive.

From a practical perspective, value conversion is also a relevant topic, and editing `vRPCConvertParams` in `rpcclient.cpp` fails the goal of modularity, so I created [json_conversion.h](https://gist.github.com/dexX7/2a38ae30074c1eade4aa), to skip conversion tables and it can be used directly on `Value` objects with `Cast<bool>(param[0])`, `Cast<int16_t>(param[1])`, ...

jonasschnelli commented:
@dexX7 Indeed. The `vRPCConvertParams` needs also modularization. I'll check your implementation approach and will try to update this PR.

jonasschnelli commented:
@dexX7 i played around with your `json_conversion.h` implementation and it seems to be useful. How would you solution to the call `RPCTypeCheck` look like? I assume when skipping the `vRPCConvertParams` conversion the type check no longer works?

We could get rid of the conversion via `vRPCConvertParams` but this would generate some changes (problem of testing/reviewing). Maybe the `json_conversion.h` should only be used for new calls.

dexX7 committed:
[Modularization] allow to add entries to the JSON conversion table

- add method to insert custom conversion entries

Conflicts:
	src/test/rpc_tests.cpp

dexX7 commented:
Good point, I didn't really intend to replace the old system, but it was more like a quick workaround-ish solution, although I have to admit, I don't like having a conversion table *and* still be required specify data types (e.g. `get_str()`, `get_int()`, ...).

I pushed three more conservative and similar, but slightly different each, approaches:

- [modularization-json-conversion-3](https://github.com/dexX7/bitcoin/commit/4aee62ad96232074685df1d41cb33a1baec746c6): one function to add new entries
- [modularization-json-conversion-2](https://github.com/dexX7/bitcoin/commit/c614786f176d9a9dcf83bc1b960d153dbf29f56b): expose conversion class and table without param class
- [modularization-json-conversion-1](https://github.com/dexX7/bitcoin/commit/9eafafd3f092ba78544afc976aecd4ec9828d219): expose conversion classes and table



As you may see, I'm a bit lost and undecided, design wise, but I hope it helps to push this further. :)

dexX7 committed:
[Modularization] use pointers when adding or replacing RPC commands

Using pointers signals the responsibility of managing object lifetime is in the hand of the caller.

jonasschnelli committed:
[Modularization] flexible RPC table overhaul

jonasschnelli commented:
pulled in @dexX7 conversion table work. Fixed code nits mentioned by @sipa .

A flexible RPC interface is required if we like to add capability to switch on/off certain modules.
It's also totally required to factor out/decouple the wallet from the rest of the code.
It could be a intermediate stop in case of a new RPC implementation. But for now i think this is a feasible way.
Reviewing this means helping bitcoind to get a new wallet-implenentation while a legacy wallet could still be enabled.

dexX7 commented:
Adding new commands is still an issue for me, if the scope is left. For example:

jonasschnelli commented:
@dexX7 Thanks for the response. Yes. Bad memory management. Will have a look at it.

jonasschnelli commented:
@dexX7 i reviewed you gist https://gist.github.com/dexX7/32accc4b04cb2e7dd496. I would say the PR is acting correct.
When adding a rpc command over the new call `AddOrReplaceCommand()` you have to make sure that your `CRPCCommand` struct stays in memory. In your gist you create your CRPCCommand within your function `void addTestCommand()` but then test it in another block. `const CRPCCommand newCmd` get freed after exiting `void addTestCommand()`.

I think one need to take the same approach to add commands to the `vRPCCommands[]` table as we do it for the standard rpc calls: `static const CRPCCommand vRPCCommands[] = ...`.

This is not perfect but would give a RPC flexibility with a diff-size of +82/−2.

dexX7 commented on src/rpcserver.h:
```diff
@@ -114,6 +114,12 @@ class CRPCTable
+     * add or replace a CRPCCommand to the dispatch table
+     * @param command rpc command to add or replace
+     */ 
+    void AddOrReplaceCommand(const CRPCCommand command);
```
The devil is in the details here, I think. :) `const CRPCCommand command` creates *copy* of the original command, and the caller can maintain the original for the whole execution lifetime, but it would still break (try: the [updated gist](https://gist.github.com/dexX7/32accc4b04cb2e7dd496)).

It's a slightly different story with `RPCAddConversion(const std::string& method, int idx)`, because the lifetime the method name seems to be extended, where [Temporary object lifetime, note 1](http://en.cppreference.com/w/cpp/language/lifetime) appears to apply to my understanding.

Upfront: using `const CRPCCommand& command` works, but the caller must still take care of the lifetime, which is not very intuitive imho, as one might be temped to believe it's behavior is similar to the `const std::string& s` case, thus my comment about passing a pointer `const CRPCCommand*` instead, which is makes it very clear that it's the caller's responsibility (in code: https://github.com/dexX7/bitcoin/commit/57b252e4c7665542dd6ff92b7e9aafab77d8811a).

Messing around with object lifetime and pointers is a bit cumbersome and error prone in general in my opinion, so it might be worth to overthink the whole approach of storing pointers in `mapCommands`, but on the other hand, keeping changes minimal and doing it step by step probably yields results more quickly.

jonasschnelli replied:
You are totally right. It must be `const CRPCCommand& command` (passing pointers). I just pulled in your commit.
I'm also aware that this pointer-passing-approche is not intuitive. But i'm still trying to keep the changeset-size at minimum. I also think a rpc calls will be added when the app starts (modules get loaded) then they stay alive until the app detects a shutdown so the lifetime management tend to be not so complex.

sipa commented:
Needs rebase.

jonasschnelli commented:
@sipa: i think this solution is more flexible (and also less risky): https://github.com/bitcoin/bitcoin/pull/6006. Closing this.
